Accessing parameters loaded from YAML file

jkaniuka opened the pull request:
This PR extends C++ and Python tutorials regarding **Using parameters in a class** with examples of accessing parameters loaded from a YAML file. This topic was discussed in https://github.com/ros2/ros2_documentation/issues/4927. I've also added a hint in [Migrating Parameters](https://docs.ros.org/en/rolling/How-To-Guides/Migrating-from-ROS1/Migrating-Parameters.html) guide regarding syntax (in ROS 2 we use dots instead of slashes while accessing parameters defined under namespaces). 

Resolves https://github.com/ros2/ros2_documentation/issues/4927

jkaniuka committed:
docs(migrating_parameters): added hint regarding usage of dots instead of slashes while accessing params in ROS 2

jkaniuka committed:
docs(python_params_in_class) : accessing params from YAML file

jkaniuka committed:
docs(cpp_params_in_class) : accessing params from YAML file

fujitatomoya commented on source/Tutorials/Beginner-Client-Libraries/Using-Parameters-In-A-Class-CPP.rst:
```diff
@@ -428,6 +428,144 @@ The terminal should return the following message the first time:
+3.4 Change via passing YAML file as an argument at node startup
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As a reminder from the :doc:`tutoral about parameters <../Beginner-CLI-Tools/Understanding-ROS2-Parameters/Understanding-ROS2-Parameters>`, you can also load parameter file at node startup.
```
i would use `:ref:` to directly point to the `ros2 param load` section as reference to guide the user.

jkaniuka replied:
Replaced `:doc:` with `:ref:` in https://github.com/jkaniuka/ros2_documentation/commit/da0a518345e1af9304d0c673c448d86f55d79cc5

fujitatomoya commented on source/Tutorials/Beginner-Client-Libraries/Using-Parameters-In-A-Class-Python.rst:
```diff
@@ -434,6 +434,148 @@ The terminal should return the following message the first time:
+3.4 Change via passing YAML file as an argument at node startup
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As a reminder from the :doc:`tutoral about parameters <../Beginner-CLI-Tools/Understanding-ROS2-Parameters/Understanding-ROS2-Parameters>`, you can also load parameter file at node startup.
```
same here, can we use `:ref:` instead?

jkaniuka replied:
Replaced `:doc:` with `:ref:` in https://github.com/jkaniuka/ros2_documentation/commit/da0a518345e1af9304d0c673c448d86f55d79cc5

fujitatomoya commented on source/How-To-Guides/Migrating-from-ROS1/Migrating-Parameters.rst:
```diff
@@ -60,6 +60,13 @@ We would construct our ROS 2 parameters file as follows:
+
+Let's say we want to use ``lidar_name`` parameter inside C++/Python node.
+In ROS 1, we used slashes to separate node name and namespaces - ``"lidar_ns/lidar_node_name/lidar_name"``.
+In ROS 2, we use dots instead of slashes - ``"lidar_ns.lidar_node_name.lidar_name"``.
```
not only adding this migration doc, but also to `Understanding parameters` section about this? i believe this is basic and primitive information about ROS 2 parameter, so it would be added to the basic guidance?

jkaniuka replied:
In [Understanding parameters](https://docs.ros.org/en/rolling/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Parameters/Understanding-ROS2-Parameters.html) tutorial, only CLI is used, and we don't access parameters from code there. I think that [Using parameters in a class](https://docs.ros.org/en/rolling/Tutorials/Beginner-Client-Libraries/Using-Parameters-In-A-Class-Python.html) is a better place to highlight this issue :smiley:

fujitatomoya replied:
this is not only for APIs but also CLI is related, `parameter_events.publisher.depth` is shown during the `Understanding parameters`. besides that, this is basic information for user, i think we can add it there too.

jkaniuka replied:
Added in https://github.com/ros2/ros2_documentation/pull/4929/commits/95d9e071f7425b02c6bd01ad2f4d2199038073c2

fujitatomoya commented on source/Tutorials/Beginner-Client-Libraries/Using-Parameters-In-A-Class-CPP.rst:
```diff
@@ -428,6 +428,144 @@ The terminal should return the following message the first time:
+
+Further outputs should show  ``[INFO] [minimal_param_node]: Hello world!`` every second.
+
+Parameter listed in ``cpp_parameters_config.yaml`` file will be set only for ``custom_minimal_param_node`` node.
+If you want to indicate that the parameter ``my_parameter`` should be set on any node in any namespace, then you should use wildcards (``/**``).
+
+.. code-block:: yaml
+
+  /**:
+    ros__parameters:
+        my_parameter: "earth"
+
+You can also use your parameter under namespace, but it will require some changes in ``cpp_parameters_node.cpp`` file.
+First, edit config file to look like the one below.
+
+.. code-block:: yaml
+
+  custom_minimal_param_node:
+    my_namespace:
+      ros__parameters:
+          my_parameter: "earth"
```
i think this is redundant with following python tutorial and already explained in https://docs.ros.org/en/rolling/Tutorials/Intermediate/Launch/Using-ROS2-Launch-For-Large-Projects.html#using-wildcards-in-yaml-files, so i would remove this explanation.

kscottz replied:
Perhaps a cross reference would be preferred here. I was looking at the current state of the parameter tutorials and we don't really push users to the intermediate and advanced topics. 

If you are a naive user you hit this tutorial first, fail to find the information you want, and are then forced to comb through the index to find what you want. It isn't obvious to me that I should look under launch to find info about using parameters. 

I think what we want here is to articulate the four possible uses, touch upon  the last two briefly, and the direct the user to where they can find more information.

jkaniuka replied:
@kscottz I agree completely :+1: . I think it's worth adding **Related content** section at the end of **Using parameters in a class** tutorials.

jkaniuka replied:
In https://github.com/jkaniuka/ros2_documentation/commit/5cc00bfe2d2e04c06ec946d7e6680be87b6521e6 I've added **Related content** sections at the end of **Using parameters in a class** tutorials. They point towards two crucial intermediate-level tutorials:
* Parameters section of [Managing large projects](https://docs.ros.org/en/rolling/Tutorials/Intermediate/Launch/Using-ROS2-Launch-For-Large-Projects.html#parameters) tutorial
* [Monitoring for parameter changes (Python/C++)](https://docs.ros.org/en/rolling/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.html) tutorial

fujitatomoya commented on source/Tutorials/Beginner-Client-Libraries/Using-Parameters-In-A-Class-CPP.rst:
```diff
@@ -428,6 +428,144 @@ The terminal should return the following message the first time:
 
 Further outputs should show  ``[INFO] [minimal_param_node]: Hello world!`` every second.
 
+3.3 Change via launch file loading parameters from YAML file
```
i am not sure if we would want to introduce this in beginner's tutorial here, looks more intermediate information that is actually explained in https://docs.ros.org/en/rolling/Tutorials/Intermediate/Launch/Using-ROS2-Launch-For-Large-Projects.html#loading-parameters-from-yaml-file. if anything missing there, probably we can add more information to https://docs.ros.org/en/rolling/Tutorials/Intermediate/Launch/Using-ROS2-Launch-For-Large-Projects.html#loading-parameters-from-yaml-file instead of here?

besides this information is redundant with python one as below except `CMakeLists.txt`.

kscottz replied:
@fujitatomoya I think this one is on me. I added a comment in the issue ticket that I thought this would be the best place to put this information. 

While redundancy isn't great I think we probably want this information where people naturally look for it, and this is probably the place. We're fighting an uphill battle against bad SEO.

jkaniuka replied:
@kscottz as you suggested, I've made the description of the third and fourth method much shorter. I've added references to existing tutorials and highlighted information about using dot in form of `.. note::` :arrow_right:  https://github.com/jkaniuka/ros2_documentation/commit/8a4a58003781b69074f5c1bbcbf167e8dde13aa5

fujitatomoya reviewed (commented):
@jkaniuka thank you very much for the contribution.

i have several comments, could you share your thoughts?

kscottz reviewed (changes requested):
@jkaniuka thanks for this pull request! We really appreciate it and we appreciate your patience while we figure out how this can land best for the readers. If you want to know why we're so scared of redundancy it is because we have to go and check all of this material with each successive ROS release.

Looking at @fujitatomoya's comments I think the best way to structure this is to leverage the existing material while adding references to those documents in the intro tutorial. 

I would say the best way to do this is to articulate the four ways parameters are used in the intro tutorial but maybe hold off writing a full tutorial for each one. What we we should instead is, for the last two, briefly summarize the concept (e.g. "you can use yaml to hold your parameters") and the key points (e.g. "yaml params have dots between them), and then reference the existing material with a full tutorial.

I hope that makes sense. Happy to clarify if it doesn't.

jkaniuka committed:
replaced :doc: with :ref: in reference

jkaniuka committed:
shorter description of setting params via YAML file - added references instead

jkaniuka committed:
added directions towards intermediate tutorials

jkaniuka commented:
Thank you @kscottz & @fujitatomoya for your reviews :smiley: I've addressed all the comments and introduced relevant changes. Happy to iterate on this :+1:

jkaniuka committed:
docs(Understanding parameters): added comment regarding dots usage while separating params namespaces and name

fujitatomoya reviewed (approved):
lgtm, thanks for iterating!

@kscottz i will leave this to you, please merge this after your approval.

kscottz reviewed (approved):
This is great! Thanks for responding so well to our feedback. 

I can approve this.

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-news-for-the-week-of-january-6th-2025/41487/1
